### PR TITLE
Render wide characters correctly to avoid being cut off

### LINKED ISSUE #791: Glyphs sometimes cut off (need to draw background in single pass)
I'm getting a visual glitch which causes the right half of certain characters to be missing. Restarting alacritty fixes it.

It happens when working with han characters, which are double width. I haven't been able to figure out the exact situation when it occurs. It seems to occur most often after displaying a lot of characters, so perhaps it's due to a cache getting filled up.

In this image, the first han character on the screen is okay, and the last three are truncated.

half-characters

### PR DESCRIPTION
**Pull request by flw-cn**
The commit contains a realistically tested workaround with no performance impact.

I made a few necessary changes to the shaders in the GLSL code, fixing the
background width of the wide characters and skipping everything about spacer
handling (in fact, they shouldn't be displayed on the screen at all, either
foreground or background).

This should close #791, cc #4038.

### CODE REVIEW

File: res/text.v.glsl [hunk 1]
```diff
@@ -44,6 +51,11 @@ uniform int backgroundPass;
 
 void main()
 {
+    skipSpacer = isSpacer;
+    if (skipSpacer != 0) {
+        return;
+    }
```
Comment (chrisduerr) [inline res/text.v.glsl:57]: I don't think it makes sense to send data to the shader, just to always discard it?

File: alacritty/src/renderer/mod.rs [hunk 2]
```diff
@@ -421,6 +421,10 @@ struct InstanceData {
     bg_a: f32,
     // Flag indicating that glyph uses multiple colors, like an Emoji.
     multicolor: u8,
+    // Set to 1 if glyph is a WIDE_CHAR.
+    is_wide_char: u8,
+    // Set to 1 if glyph is a WIDE_CHAR_SPACER.
+    is_spacer: u8,
```
Comment (kchibisov) [inline alacritty/src/renderer/mod.rs:427]: I guess we can reuse multicolor value and call it like cell_flags and just pass all three values in it. Having 3 boolean flags to pass to GPU is wasteful to me.

Comment (flw-cn) [reply alacritty/src/renderer/mod.rs:427]: Ok, let me have a try.

File: res/text.v.glsl (hunk 1 above)
Comment (flw-cn) [reply res/text.v.glsl:57]: Are you saying we should just remove the spacer rendering operation from the Rust code?
Let me have a try.

Comment (chrisduerr) [reply res/text.v.glsl:57]: Yes. We are already filtering out whitespace characters. So if the spacer is not needed for rendering, we should filter it out. Though make sure to filter both the leading and trailing spacers.

**Comment by flw-cn** [comment]
I've discarded the spacer rendering action.

But when I try to reduce the flags I run into a lot of trouble.

* The smallest GLSL data types are 32-bit, but flags use `u8` to save space (if we use `u32`, it might be simpler to use four `u8` instead).
* After I pass the flags of u8, the bit is not easily and intuitively retrieved in the GLSL script, in fact, there is a problem with the bit-order. For example, if I pass `flags = 0x02`, I will receive `0x4000_0000` in GLSL.
* The FLAG_NAME/VALUE in the GLSL script must also correspond to the one in Rust, but this can easily go wrong during maintenance late on.
* Even if I'm lucky enough to do it, I don't have the energy to test all other platforms to make sure it's universally correct.

In summary, I'm not going to merge the flags, and I've kept a separate `WIDE_CHAR` flag.

If you have a better suggestion on what to do, please let me know. @chrisduerr @kchibisov

File: alacritty/src/renderer/mod.rs [hunk 3]
```diff
@@ -1023,6 +1037,12 @@ impl<'a> RenderApi<'a> {
 
     #[inline]
     fn add_render_item(&mut self, cell: RenderableCell, glyph: &Glyph) {
+        if cell.flags.contains(Flags::WIDE_CHAR_SPACER)
+            && !cell.flags.contains(Flags::LEADING_WIDE_CHAR_SPACER)
+        {
+            return;
+        }
```
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:1044]: Things should be filtered out earlier, not in the renderer.

**Comment by flw-cn** [comment]
Anyway, I did my best. I hope my fork can give you some inspiration. Others are also welcome to follow up to resolve this issue once and for all.

**Comment by chrisduerr** [comment]
Closing since @flw-cn does not seem interested in finishing up this PR.
